File: sintropy-engine/src/main/kotlin/com/ph/sintropyengine/broker/model/EventLog.kt
package com.ph.sintropyengine.broker.model

import java.time.OffsetDateTime
import java.util.UUID
import org.jooq.JSONB

data class EventLog(
    val messageId: UUID,
    val timestamp: OffsetDateTime,
    val channelId: UUID,
    val producerId: UUID,
    val routingKey: String,
    val message: JSONB,
    val headers: JSONB,
    val processed: Boolean
)

File: sintropy_engine/lib/sintropy_engine/messages/event_log.ex
defmodule SintropyEngine.Messages.EventLog do
  use Ecto.Schema
  import Ecto.Changeset

  alias SintropyEngine.Channels.Channel
  alias SintropyEngine.Producers.Producer

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id
  schema "event_log" do
    field :timestamp, :utc_datetime
    field :routing_key, :string
    field :message, :string
    field :headers, :string
    field :processed, :boolean, default: false

    belongs_to :channel, Channel
    belongs_to :producer, Producer

    timestamps(type: :utc_datetime)
  end

  @doc false
  def changeset(event_log, attrs) do
    event_log
    |> cast(attrs, [
      :message_id,
      :timestamp,
      :routing_key,
      :message,
      :headers,
      :processed,
      :channel_id,
      :producer_id
    ])
    |> validate_required([
      :message_id,
      :timestamp,
      :routing_key,
      :message,
      :headers,
      :processed,
      :channel_id,
      :producer_id
    ])
    |> foreign_key_constraint(:channel_id)
    |> foreign_key_constraint(:producer_id)
  end
end
